test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,234 @@
+    // statically - but the whole point of this test is to have Envoy dynamically discover it!
+    test_server_->server().dispatcher().post([this]() -> void {
+      EXPECT_TRUE(test_server_->server().listenerManager().addOrUpdateListener(
+          buildListener("http", ClusterName), "", true));
+    });
+    test_server_->waitForCounterGe("listener_manager.listener_create_success", 1);
+
+    registerTestServerPorts({"http"});
+    test_server_->waitUntilListenersReady();
+  }
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, CdsIntegrationTest, GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// 1) Envoy starts up with no static clusters (other than the CDS-over-gRPC server).
+// 2) Envoy is told of a cluster via CDS.
+// 3) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+// 4) Envoy is told that cluster is gone.
+// 5) We send Envoy a request, which should 503.
+// 6) Envoy is told that the cluster is back.
+// 7) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+TEST_P(CdsIntegrationTest, CdsClusterUpDownUp) {
+  // Calls our initialize(), which includes establishing a listener, route, and cluster.
+  testRouterHeaderOnlyRequestAndResponse(true, nullptr, 1);
+
+  // Tell Envoy that cluster_0 is gone.
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "1", {}));
+  sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster, {}, "42");
+  // We can continue the test once we're sure that Envoy's ClusterManager has made use of
+  // the DiscoveryResponse that says cluster_0 is gone.
+  test_server_->waitForCounterGe("cluster_manager.cluster_removed", 1);
+
+  // Now that cluster_0 is gone, the listener (with its routing to cluster_0) should 503.
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/unknown", "", downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_STREQ("503", response->headers().Status()->value().c_str());
+
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+  fake_upstream_connection_ = nullptr;
+
+  // Tell Envoy that cluster_0 is back.
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "42", {}));
+  sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster,
+                                                 {buildCluster(ClusterName)}, "413");
+
+  // We can continue the test once we're sure that Envoy's ClusterManager has made use of
+  // the DiscoveryResponse describing cluster_0 that we sent. Again, 2 includes CDS server.
+  test_server_->waitForGaugeGe("cluster_manager.active_clusters", 2);
+
+  // Does *not* call our initialize().
+  testRouterHeaderOnlyRequestAndResponse(true, nullptr, 1);
+
+  cleanupUpstreamAndDownstream();
+  fake_upstream_connection_ = nullptr;
```

Comment:
Nit: Is this strictly needed? Won't the destructor clean this up?

Reply:
Done.